third_party/amd/lib/TritonAMDGPUTransforms/StreamPipelineV2.cpp

```diff
@@ -0,0 +1,791 @@
+static bool preConditionInner(scf::ForOp forOp) {
+  // Skip loop with distance > 1 for now.
+  // TODO: relax the constraint in the expander.
+  if (llvm::any_of(forOp.getBody()->getTerminator()->getOperands(),
+                   [](Value operand) {
+                     Operation *def = operand.getDefiningOp();
+                     return !def;
+                   }))
+    return false;
+  // Don't pipeline outer loops.
+  if (forOp
+          ->walk([&](Operation *op) {
+            if (forOp.getOperation() == op)
+              return WalkResult::advance();
+            if (isa<scf::ForOp, scf::WhileOp>(op))
+              return WalkResult::interrupt();
+            return WalkResult::advance();
+          })
+          .wasInterrupted())
+    return false;
+  return true;
+}
+
+static bool pipelineLoop(scf::ForOp forOp, int numStages) {
+  mlir::triton::PipeliningOption options;
+  if (!preConditionInner(forOp))
+    return false;
+
+  bool foundSchedule = false;
+  foundSchedule = preProcessLoopAndGetSchedule2(forOp, numStages, options);
+
+  // TODO: add more pipelines strategy.
+  if (!foundSchedule)
+    return false;
+
+  IRRewriter rewriter(forOp->getContext());
+  rewriter.setInsertionPoint(forOp);
+  FailureOr<scf::ForOp> newForOp =
+      mlir::triton::pipelineForLoop(rewriter, forOp, options);
+
+  if (failed(newForOp))
+    return false;
+  return true;
+}
+
+namespace {
+struct PipelinePass : public TritonAMDGPUStreamPipelineV2Base<PipelinePass> {
+  PipelinePass() = default;
+  PipelinePass(int32_t numStages) { this->numStages = numStages; }
+
+  int getNumStagesOrDefault(scf::ForOp forOp) {
+    // Use the attribute attached to the loop if it exists otherwise use the
+    // global control.
+    if (auto attr =
+            forOp->getAttrOfType<IntegerAttr>(mlir::triton::kNumStagesAttrName))
+      return attr.getInt();
+    return numStages;
+  }
+
+  void runOnOperation() override {
+    SmallVector<scf::ForOp> loops;
+    getOperation()->walk([&](scf::ForOp forOp) {
+      // Bail out for loops with num_stage <= 1.
+      if (getNumStagesOrDefault(forOp) > 1)
+        loops.push_back(forOp);
+    });
+
+    if (loops.empty())
+      return;
+
+    bool pipelined = false;
+    for (scf::ForOp forOp : loops) {
+      auto outerLoop = dyn_cast<scf::ForOp>(forOp->getParentOp());
+      int loopNumStages = getNumStagesOrDefault(forOp);
+      pipelined |= pipelineLoop(forOp, loopNumStages);
+    }
+
+    if (pipelined) {
+      // Clean up arithmetic before applying the next level of pipelining to
+      // simplify the IR.
+      auto arithDialect =
+          getOperation().getContext()->getLoadedDialect<arith::ArithDialect>();
+      RewritePatternSet patterns(getOperation().getContext());
+      arithDialect->getCanonicalizationPatterns(patterns);
+      if (applyPatternsAndFoldGreedily(getOperation(), std::move(patterns))
+              .failed())
+        signalPassFailure();
+    }
```

Comment:
I don't think it is needed anymore. This was there to clean up before the outer loop pipelining but we will deprecate it in favor of Pawel's fused loop pipeliner.